dashboard/transform_dashboard.py

```diff
@@ -0,0 +1,102 @@
+"""Script containing the functions to create altair visualisations for streamlit dashboard."""
+
+from os import environ as ENV
+
+from dotenv import load_dotenv
+import pandas as pd
+import altair as alt
+from datetime import datetime, timedelta
+
+from extract_dashboard import get_db_connection
+
+
+load_dotenv()
+
+
+def get_botanist_mapping():
+    """Creates dictionary of botanist names and ids."""
+
+    with get_db_connection() as conn:
+
+        schema_name = ENV['SCHEMA_NAME']
+
+        query = f'''
+        SELECT botanist_id, first_name, last_name FROM {schema_name}.botanist
+        '''
+
+        df = pd.read_sql(query, conn)
+
+        id_to_name = dict(
+            zip(df['botanist_id'], df['first_name'] + ' ' + df['last_name']))
+
+        return id_to_name
+
+
+def create_botanist_pie(df: pd.DataFrame) -> alt.Chart:
+    """Creates pie chart of botanists as a proportion of how many plants they monitor."""
+
+    grouped_df = df.groupby(df["botanist_id"]).count().reset_index()
+
+    botanist_mapping = get_botanist_mapping()
+    grouped_df["botanist_name"] = grouped_df["botanist_id"].map(
+        botanist_mapping)
+
+    title = alt.TitleParams('Botanist plant count', anchor='middle')
+    botanist_pie_chart = alt.Chart(grouped_df, title=title).mark_arc().encode(
+        theta="plant_id:Q",
+        color="botanist_name:N"
+    )
+
+    return botanist_pie_chart
+
+
+def create_temperature_bar(df: pd.DataFrame):
+    """Creates bar chart of plants against their most recently recorded temperature."""
+
+    df['at'] = pd.to_datetime(df['at'])
+
+    recent_df = df.loc[df.groupby('plant_id')['at'].idxmax()]
+
+    title = alt.TitleParams('Plant temperature', anchor='middle')
+    temp_bar_chart = alt.Chart(recent_df, title=title).mark_bar().encode(
+        x='plant_id:N',
+        y='temperature:Q'
+    )
+
+    return temp_bar_chart
+
+
+def create_temperature_line(df: pd.DataFrame):
+    """Creates bar chart of plants against their most recently recorded temperature."""
+
+    df['at'] = pd.to_datetime(df['at'])
+
+    three_hours_ago = df['at'].max() - pd.Timedelta(hours=3)
+
+    # Filter the DataFrame to only include readings from the past hour
+    df_filtered = df[df['at'] >= three_hours_ago]
+
+    title = alt.TitleParams('Plant temperature over time', anchor='middle')
+    temp_line_chart = alt.Chart(df_filtered, title=title).mark_line().encode(
+        x='at:T',  # 'at' column for time on x-axis
+        y='temperature:Q',  # 'temperature' column for y-axis (quantitative)
+        color='plant_id:N'  # Different color for each plant_id
+    )
+
+    return temp_line_chart
+
+
+def create_moisture_bar(df: pd.DataFrame):
```

Comment:
I like these functions for creating the functions. They could be type hinted with `alt.Chart` as their return value, though!